File: core/storage/interface.py
from abc import ABC, abstractmethod
from typing import Optional, BinaryIO


class FileStorageInterface(ABC):

    @abstractmethod
    def get_default_bucket(self) -> str:
        ...

    @abstractmethod
    def store(self, bucket: str, file_key: str, data: bytes) -> str:
        ...

    @abstractmethod
    def store_stream(self, bucket: str, file_key: str, stream: BinaryIO) -> str:
        ...

    @abstractmethod
    def get_bytes(self, bucket: str, file_key: str) -> bytes:
        ...

    @abstractmethod
    def get_url(self, bucket: str, file_key: str) -> str:
        ...

    @abstractmethod
    def get_auth_url(self, bucket: str, file_key: str, timeout_ms: int = 60000) -> str:
        ...

    @abstractmethod
    def delete(self, bucket: str, file_key: str) -> None:
        ...

    @abstractmethod
    def exists(self, bucket: str, file_key: str) -> bool:
        ...

    @abstractmethod
    def copy(self, src_bucket: str, src_key: str, dst_bucket: str, dst_key: str) -> None:
        ...

File: core/storage/interface.go
package storage

import "io"

// FileStorageInterface defines the contract for file storage backends.
// Follows the same abstraction as hei-fastapi's storage interface.
type FileStorageInterface interface {
	// GetDefaultBucket returns the default bucket/namespace name.
	GetDefaultBucket() string

	// Store stores raw bytes under the given key in the specified bucket.
	Store(bucket, fileKey string, data []byte) (string, error)

	// StoreStream stores data from a reader under the given key.
	StoreStream(bucket, fileKey string, reader io.Reader) (string, error)

	// GetBytes retrieves the raw bytes for a given key.
	GetBytes(bucket, fileKey string) ([]byte, error)

	// GetURL returns the direct access URL for the given key.
	GetURL(bucket, fileKey string) (string, error)

	// GetAuthURL returns a time-limited authenticated URL.
	GetAuthURL(bucket, fileKey string, timeoutMs int) (string, error)

	// Delete removes the object at the given key.
	Delete(bucket, fileKey string) error

	// Exists checks if an object exists at the given key.
	Exists(bucket, fileKey string) (bool, error)

	// Copy copies an object from source to destination.
	Copy(srcBucket, srcKey, dstBucket, dstKey string) error
}
